### PR TITLE
[NEAT-27] DMS sheet validation

### PR DESCRIPTION
**Pull request by doctrino**
Validations added:

1. Container properties are defined are referenced over multiple rows, this ensures that the container definition is consistent.
2. Check of the value type is valid.

In addition, allow you to only define index and constraint on a container property only the first row you define it. You can set the rest to zero.

Instead of 

image

You can now do:
image

### CODE REVIEW

File: cognite/neat/rules/models/_rules/_types.py [hunk 1]
```diff
@@ -114,12 +114,20 @@ def _check_parent(value: list[ParentClass]) -> list[ParentClass]:
...
+        return_type=str,
+        when_used="unless-none",
+    ),
 ]
 
 
 StrListType = Annotated[
     list[str],
     BeforeValidator(lambda value: [entry.strip() for entry in value.split(",")] if isinstance(value, str) else value),
+    PlainSerializer(
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/_types.py:128]: Is this used when exporting rules to e.g. Excel?

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 2]
```diff
@@ -221,6 +233,75 @@ class DMSRules(BaseRules):
...
+                container_properties_by_id[(prop.container, prop.container_property)].append(prop)
+
+        exceptions: list[str] = []
+        for (container, prop_name), properties in container_properties_by_id.items():
+            if len(properties) == 1:
+                continue
+
+            value_types = {prop.value_type for prop in properties if prop.value_type}
+            if len(value_types) > 1:
+                exceptions.append(
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:250]: Define these as NeatExceptions

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 3]
```diff
@@ -221,6 +233,75 @@ class DMSRules(BaseRules):
...
+                    f"Container {container}.{prop_name} is defined with different value types: {value_types}"
+                )
+            list_definitions = {prop.is_list for prop in properties if prop.is_list is not None}
+            if len(list_definitions) > 1:
+                exceptions.append(
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:255]: Same here

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 4]
```diff
@@ -221,6 +233,75 @@ class DMSRules(BaseRules):
...
+                    f"Container {container}.{prop_name} is defined with different "
+                    f"list definitions: {list_definitions}"
+                )
+            nullable_definitions = {prop.nullable for prop in properties if prop.nullable is not None}
+            if len(nullable_definitions) > 1:
+                exceptions.append(
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:261]: same here

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 5]
```diff
@@ -221,6 +233,75 @@ class DMSRules(BaseRules):
...
+                    f"Container {container}.{prop_name} is defined with different "
+                    f"nullable definitions: {nullable_definitions}"
+                )
+            default_definitions = {prop.default for prop in properties if prop.default is not None}
+            if len(default_definitions) > 1:
+                exceptions.append(
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:267]: same here

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 6]
```diff
@@ -221,6 +233,75 @@ class DMSRules(BaseRules):
...
+                    f"Container {container}.{prop_name} is defined with different "
+                    f"default definitions: {default_definitions}"
+                )
+            index_definitions = {",".join(prop.index) for prop in properties if prop.index is not None}
+            if len(index_definitions) > 1:
+                exceptions.append(
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:273]: same

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 7]
```diff
@@ -221,6 +233,75 @@ class DMSRules(BaseRules):
...
+                    f"Container {container}.{prop_name} is defined with different "
+                    f"index definitions: {index_definitions}"
+                )
+            constraint_definitions = {",".join(prop.constraint) for prop in properties if prop.constraint is not None}
+            if len(constraint_definitions) > 1:
+                exceptions.append(
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:279]: same as aboce

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 8]
```diff
@@ -221,6 +233,75 @@ class DMSRules(BaseRules):
...
+                prop.value_type = value_type
+                prop.is_list = prop.is_list or list_definition
+                prop.nullable = prop.nullable or nullable_definition
+                prop.default = prop.default or default_definition
+                prop.index = prop.index or index_definition
+                prop.constraint = prop.constraint or constraint_definition
+
+        if exceptions:
+            exception_str = "\n".join(exceptions)
+            raise ValueError(f"Inconsistent container(s): {exception_str}")
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:302]: You can actually use `ExceptionGroup` for this, 

For python lower than 3.11, make sure to import this:

```
if sys.version_info < (3, 11):
    from exceptiongroup import ExceptionGroup
```

**Review by nikokaoja — APPROVED** [review]
Couple of comments regarding exception handling.

File: cognite/neat/rules/models/_rules/_types.py (hunk 1 above)
Comment (doctrino) [reply cognite/neat/rules/models/_rules/_types.py:128]: It is used when you call `.model_dump`

File: cognite/neat/rules/models/_rules/dms_architect_rules.py (hunk 8 above)
Comment (doctrino) [reply cognite/neat/rules/models/_rules/dms_architect_rules.py:302]: I tried, but pydantic was not happy.

**Pull request merged**